src/dbcp/transform/acp_projects.py

```diff
@@ -0,0 +1,458 @@
+    assert (
+        raw_df["MW_Total_Capacity"].dtype == pd.Float64Dtype()
+    ), "Capacity is not float dtype"
+
+    to_hash = raw_df.loc[:, pk].copy()
+    str_cols = to_hash.select_dtypes(include="string").columns
+    to_hash[str_cols] = to_hash[str_cols].apply(
+        lambda s: s.str.strip().str.lower(), axis=1
+    )
+    # Protect against floating point representation shenanigans by printing capacity to
+    # 3 decimal places (the max observed precision in the data as of 2024-07-28)
+    to_hash["MW_Total_Capacity"] = to_hash["MW_Total_Capacity"].apply(
+        lambda num: f"{num:.3f}"
+    )
+    concat = to_hash.fillna("").apply(" ".join, axis=1)
+    out = concat.apply(_int_id_from_str)
+    assert out.is_unique, "Surrogate key is not unique"
+    return out
+
+
+def _int_id_from_str(s: str) -> int:
+    """Use a hash function to convert a string to a unique integer."""
+    # hash with md5, subset 16 byte digest to first 8 bytes and cast to int64 because
+    # integers are more convenient and we don't need the full 16 bytes for data this
+    # small (collision probability is on the order of 1e-12 for 10k items with 64 bits)
+    byte_str = s.encode("utf-8")
+    hash_digest = hashlib.md5(byte_str).digest()[:8]  # nosec
+    # Specify byteorder and signedness just for clarity. I don't *really* care what they
+    # are as long as the underlying bytes are unique. But Postgres has no unsigned
+    # integer types, so I'll use signed int for consistency between the pre-db and
+    # post-db data.
+    int_id = int.from_bytes(hash_digest, byteorder="little", signed=True)
+    return int_id
+
+
+def transform(raw_dfs: dict[str, pd.DataFrame]) -> dict[str, pd.DataFrame]:
+    """Transform and clean ACP projects data."""
+    # rename columns
+    trans = raw_dfs["raw_acp_projects"].convert_dtypes()
+    surrogate_key = _make_surrogate_key(trans)  # uses raw column names
+    trans.columns = _rename_columns(trans.columns)
+    trans["proj_id"] = surrogate_key  # assign after renaming columns
+
+    def _str_strip(ser):
+        return ser.str.strip()
+
+    # clean priority columns. Leave the rest as raw_
+    col_transforms: dict[str, Callable] = {
+        "raw_status": _col_transform_status,
+        "raw_eia_plant_code": _col_transform_eia_plant_code,
+        "raw_project_name": _str_strip,
+        "raw_phase_name": _str_strip,
+        "raw_phase_type": _col_transform_phase_type,
+        "raw_developers": _str_strip,
+        "raw_owners": _str_strip,
+        "raw_iso_rtos": _col_transform_iso_rtos,
+        "raw_owner_types": _col_transform_owner_types,
+        "raw_mw_total_capacity": partial(
+            _col_transform_mw_total_capacity, full_df=trans
+        ),
+    }
+    for raw_col, transform in col_transforms.items():
+        new_col_name = raw_col[4:]  # remove 'raw_'
+        trans[new_col_name] = transform(trans[raw_col])
+    county_shapes = _extract_census_counties(CENSUS_URI).set_geometry("geometry")
```

Comment:
It'd be nice to use dagster here so this acp asset could depend on the raw census asset.

Reply:
Yes it would!